Define initial Swagger API for CDR index

calbach opened the pull request:
From previous design discussion, the resource hierarchy is as follows:

- Datasets (access control later will be implemented here)
  - Individuals (participants + labels)
  - DataPointers (URIs, e.g. to files/tables + labels)

Individuals and DataPointers have a many:many relationship with one another.

mikeas1 commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+        },
+        "uri": {
+          "type": "string",
+          "description": "Unique within a dataset."
```
So I get that this is the only thing that can really be said about this field since there's really no validation (I expect, but could be wrong) that this is actually a URI, but even still you might want to expand this description. Maybe something like "A URI for the resource this DataPointer references. ..add examples for common URIs... A URI must be unique within a dataset."

calbach replied:
Add more here.

mikeas1 commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+          "items": {
+            "type": "string"
+          },
+          "description": "One-of individualNames or allIndividuals; full individual names of the form 'datasets/*/individuals/*'."
```
Making individual IDs full resource names implies you can reference individuals in other datasets. Is that valid?

calbach replied:
That's not valid, made that explicit in the docs. I'm using names here for API consistency (I'd like to use names throughout), though I do see that it's allowing an unnecessary degree of freedom.

mikeas1 commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+    "CdrDataset": {
+      "type": "object",
+      "properties": {
+        "name": {
```
nit: Description for this? Can't tell if this is server-generated or something the user can set. Also the name (referenced in the schema) vs. id (referenced in paths) terminology is confusing throughout.

calbach replied:
Added a doc to the top level to explain this (hopefully).

mikeas1 commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+        ]
+      }
+    },
+    "/v1/datasets/{datasetId}/dataPointers": {
```
No search functionality? Is that because searches should be executed against ElasticSearch directly?

calbach replied:
For now, yes. Once we've scoped down the functionality we'll explicitly define that here and disallow direct ElasticSearch access.

calbach commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+            "type": "string"
+          }
+        ],
+        "tags": [
```
Yeah, this was something that the grpc-gateway (my initial seed code for this) produced. It's a bit odd. The effect of this is that the generated code will be grouped differently, and the Swagger UI will group the methods into subsections.

I switched this to group by resource type, since I think that's the more logical grouping across the board.

calbach commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+        ]
+      }
+    },
+    "/v1/datasets/{datasetId}/dataPointers/{dataPointerId}": {
```
Agreed, done.

calbach commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+      "delete": {
+        "operationId": "DeleteDataPointer",
+        "responses": {
+          "200": {
```
Yeah I'm also not too sure about convention. I think the main goal of documenting this is to give the client some idea of what JSON they can expect in the response for a given status code. Lets not bother defining that just yet, but I'll leave that as a future extension and just document the initial set of failure codes.

I've also omitted 403 here since we won't have that for the initial version, but it should be added later.

calbach commented on src/index.swagger.json:
```diff
@@ -0,0 +1,462 @@
+      "properties": {
+        "name": {
+          "type": "string",
+          "description": "Server generated."
```
Added documentation. Anything "id" is a resource ID, e.g. "y" in "datasets/x/individuals/y". Anything "name" is the whole thing: "datasets/x/individuals/y".

I added some cross-cutting explanation of this to the top-level API doc for lack of a better home.

calbach commented:
Thanks for the review. In the future I think I'll use follow-up commits to respond to comments which should hopefully improve the reviewer diff + comment flow - then I'll squash at the last second.

FYI: I made two more minor changes after your approval:
- Removed "required: False" for the pagination params, since those are the default setting
- Dropped the 'Cdr' prefix from the resource definitions. These don't seem necessary.

@davidangb ready for you to take a look. Thanks.

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,426 @@
+produces:
+  - application/json
+paths:
+  /v1/datasets:
```
consider using a swagger `basePath: /v1` and excluding `v1` from each path.

calbach replied:
Done.

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,426 @@
+        '500':
+          $ref: '#/responses/ServerError'
+      parameters:
+        - name: datasetId
```
the `datasetId` path parameter is used in multiple endpoints. Consider moving it to a top-level `parameters` section and reusing it via `$ref`s

Same goes for `dataPointerId` and `individualId` later in the document.

calbach replied:
Done.

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,426 @@
+      name:
+        description: |
+          Unique client-specified resource name of the form 'datasets/*'.
+        type: string
```
do you care to add `required` constraints into the swagger definitions at this point?

calbach replied:
Added, this is actually the only place it belongs, I think.

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,426 @@
+        $ref: '#/definitions/labels'
+  labels:
+    type: object
+    additionalProperties: true
```
syntax nitpick: as a top-level definition, I would capitalize `Labels`

calbach replied:
Done.

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,426 @@
+  /v1/datasets:
+    get:
+      description: |
+        Paginated method for listing all datasets in the repository.
```
for the list api(s), if we do not allow the end user to sort (yet), we should at least say something in the doc about the sort order, even if we say the sort order is arbitrary but deterministic.

calbach replied:
Good point, done

calbach commented:
Thanks for the review. I know I said I'd do a commit this time to improve the reviewer experience but I decided to just squash instead to ensure I'm ready for merging since you already +1'd.

> Higher-level question: should we have list apis for individuals and data pointers? i.e. "get me all the individuals in a dataset" at GET /v1/datasets/{datasetId}/individuals and "get me all the data pointers in a dataset" at GET /v1/datasets/{datasetId}/dataPointers?

Yes, we will eventually have these methods defined. We're still figuring out exactly what subset of functionality we want to expose. For the short-term, we'll just hit the ElasticSearch API directly, which gives us the full breadth of query functionality.

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,395 @@
+          Unique client-specified resource name of the individual of the form
+          'datasets/*/individuals/*'.
+      labels:
+        $ref: '#/definitions/Labels'
```
inside a schema definition, the `required` syntax is a bit different. Try:
```
  Individual:
    description: |
      Individuals, for example particpant or organisms, within a dataset.
    type: object
    required:
      - name
    properties:
      name:
        type: string
        description: |
          Unique client-specified resource name of the individual of the form
          'datasets/*/individuals/*'.
      labels:
        $ref: '#/definitions/Labels'
```

this applies in a few places under the `definitions` key

calbach replied:
Thanks for catching this. I'd been using swagger codegen to check my changes, but I guess that doesn't actually validate the syntax, will look into a linter tool..

davidangb commented on api/index.swagger.yaml:
```diff
@@ -0,0 +1,395 @@
+    required: true
+    type: string
+  DataPointerPathId:
+    name: datasetId
```
`dataPointerId`, not `datasetId`

calbach replied:
Oops, don.
